### PR TITLE
Persistent UDF Materialization

### LINKED ISSUE #451: [CT-1732] [Feature] Add Native Persistent UDF Materialization
### Is this your first time submitting a feature request?

- [X] I have read the [expectations for open source contributors](https://docs.getdbt.com/docs/contributing/oss-expectations)
- [X] I have searched the existing issues, and I could not find an existing issue for this feature
- [X] I am requesting a straightforward extension of existing dbt-bigquery functionality, rather than a Big Idea better suited to a discussion

### Describe the feature

There should be a native materialization for persistent UDFs in BigQuery. This materialization will enable users to create UDFs via `dbt run` and to track them appropriately within their DAGs and lineage graphs.

Previously, BigQuery only supported temporary UDFs. The community wrote several issues related to the handling of temporary BigQuery UDFs. For example:

- https://github.com/dbt-labs/dbt-core/issues/1879

- https://github.com/dbt-labs/dbt-core/issues/1112

However, BigQuery now supports persistent UDFs. We should have a native materialization for persistent BigQuery UDFs! This issue - https://github.com/dbt-labs/dbt-core/issues/136 - was created a while ago for cross-database UDF handling. The issue at hand, however, specifically concerns persistent BigQuery UDFs.

### Describe alternatives you've considered

A few popular alternatives discussed throughout the community:

- Create UDFs via hooks/run-operations. For example, as documented [here](https://discourse.getdbt.com/t/using-dbt-to-manage-user-defined-functions/18/1)
- Maintain UDF SQL files in the `models` directory with the `CREATE OR REPLACE` statement in the SQL. Use `dbt` to compile the model SQL and reference the UDF models via `ref`. This way, UDF models show up in the lineage graph. However, UDF creation is handled outside of `dbt run`

### Who will this benefit?

`dbt` users who leverage persistent BigQuery UDFs and want to track them in their DAGs and lineage graphs. This is particularly useful for those who leverage UDFs to define reusable, core business logic.

### Are you interested in contributing this feature?

Yes! I've started working on this. Interested to see what people think.

### Anything else?

Some questions:

- Does it make sense to implement this for BigQuery separately? I know there's interest in supporting a UDF materialization across different databases, but that seems like a much heavier lift to do all at once.
- Should this touch on temporary UDFs as well? Perhaps a UDF materialization that also supports the notion of an "ephemeral" (i.e., temporary) UDF?

### PR DESCRIPTION
**Pull request by anaghshineh**
resolves #451 

### Description

Adds `udf` materialization in support of BQ persistent SQL UDFs. This materialization allows users to declare SQL UDFs as models and manage them using `dbt` commands, like `dbt ls`, `dbt compile`, and `dbt run`.

The new materialization takes two optional configuration arguments - `args` and `return_type`. The former is an array of dictionaries. Each individual dictionary represents a single UDF argument. A single dictionary consists of `name` and `type` keys. The `name` key specifies the name of the argument. The `type` key specifies the type of the argument (e.g.,. `INT64` or `STRING`). An array is used to preserve the order of the arguments provided - the order in which arguments are listed in the array will be the order in which they are declared as arguments for the associated UDF. The `return_type` configuration argument specifies the type of the item returned by the UDF (e.g., `STRING` or `STRUCT<domain STRING, path STRING>`).

### Checklist

- [x] I have read [the contributing guide](https://github.com/dbt-labs/dbt-bigquery/blob/main/CONTRIBUTING.md) and understand what's expected of me
- [x] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
- [x] I have run this code in development and it appears to resolve the stated issue
- [x] This PR includes tests, or tests are not required/relevant for this PR
- [ ] I have [opened an issue to add/update docs](https://github.com/dbt-labs/docs.getdbt.com/issues/new/choose), or docs changes are not required/relevant for this PR
- [x] I have run `changie new` to [create a changelog entry](https://github.com/dbt-labs/dbt-bigquery/blob/main/CONTRIBUTING.md#Adding-CHANGELOG-Entry)

### CODE REVIEW

File: dbt/adapters/bigquery/connections.py [hunk 1]
```diff
@@ -587,6 +587,20 @@ def table_ref(database, schema, table_name):
...
+        return google.cloud.bigquery.RoutineReference.from_string(f"{database}.{schema}.{routine_name}")
+
+    def get_bq_routine(self, database: str, schema: str, identifier: str) -> google.cloud.bigquery.Routine:
+        """Get a BigQuery routine (UDF) for a schema/model."""
+        conn = self.get_thread_connection()
+        # backwards compatibility: fill in with defaults if not specified
+        database = database or conn.credentials.database
+        schema = schema or conn.credentials.schema
+        routine_ref = self.routine_ref(database, schema, identifier)
+        return conn.handle.get_routine(routine_ref)
```
Comment (dataders) [inline dbt/adapters/bigquery/connections.py:602]: any reason you can't use `info_schema.routines`? If we convert these python functions to SQL macro equivalents, we could publish this materialization as a dbt package (at least for an trial period).
https://cloud.google.com/bigquery/docs/information-schema-routines

**Review by dataders — COMMENTED** [review]
hey @anaghshineh this is so cool! Before we think about merging this code in, I'd love to share this with the community. Would you be interesting in forking jaffle shop to show how this new materialization would be used? Maybe even recording a quick demo showing how the DAG changes?

another random thought. what happens if you `{{ ref() }}` a UDF from another model again?

**Comment by jtcohen6** [comment]
@anaghshineh This is awesome, thanks so much for taking the time & initiative to contribute. I haven't had a chance to play around with this yet, but to @dataders' point, it might already be in a place where community members could copy-paste macros (or install from a package!) to try this out and give early feedback.

I left a big comment + question over on the linked issue: https://github.com/dbt-labs/dbt-bigquery/issues/451#issuecomment-1372164781

**Comment by Fleid** [comment]
Hi @anaghshineh, I'm going to close that PR for now.

We haven't reach a clear consensus on what to do with UDFs yet, and next step would be to build momentum towards that following @dataders' [advice](https://github.com/dbt-labs/dbt-bigquery/pull/454#pullrequestreview-1236498047).
Please let us know if you need help getting there, if you do want to keep pushing on that topic.

This is fantastic work that you did here, and I feel bad about closing this PR.
Now that comes with the territory of dbt being an opinionated tool - @jtcohen6 [explained in details](https://github.com/dbt-labs/dbt-bigquery/issues/451) the context on this specific issue - which we all love, but still, it's not a great feeling.

We will leverage your work if/when the situation evolves.

**Comment by brabster** [comment]
I think this is an independent experimental implementation in a real-world public project https://tempered.works/posts/2024-02-19-udf-dbt-models/ (mentioned on Slack channel #i-made-this)

**Comment by brabster** [comment]
Note discussion #10395 on support for UDFs as a materialization

**Pull request closed without merging**